src/moi.jl

```diff
@@ -406,13 +413,186 @@ function build_bilevel(
         pass_names(m, lower_dual, lower_dual_idxmap)
     end
 
+    #=
+        Check if Upper bilinear terms of the form Upper primal * Lower dual exist, and can be linearized
+    =#
+    # TODO implement for bilinear products in constraints (only checking upper objective so far)
+
+    if linearize_bilinear_upper_terms 
+        if MOI.get(upper, MOI.ObjectiveFunctionType()) <: MOI.ScalarQuadraticFunction &&
+            !isempty(upper_var_to_lower_ctr)
+            
+            linearize = true
+            # check lower constraint types and if not just equality and singlevariable bounds then linearize = false and @warn
+            lower_con_types = Set(MOI.get(lower, MOI.ListOfConstraints()))
+            standard_form_con_types = Set([
+                (MOI.ScalarAffineFunction{Float64}, MOI.EqualTo{Float64}),
+                (MOI.SingleVariable, MOI.GreaterThan{Float64}),
+                (MOI.SingleVariable, MOI.LessThan{Float64})
+            ])
+
+            if !(issubset(lower_con_types, standard_form_con_types))
+                linearize = false
+                @warn("The lower model must be in standard form to linearize bilinear terms, i.e. the constraint types must be a subset of $(standard_form_con_types). Skipping linearization process.")
+            else
+                A_N, bilinear_upper_dual_to_quad_term, bilinear_upper_dual_to_lower_primal, lower_primal_var_to_lower_con = 
+                    check_upper_objective_for_bilinear_linearization(upper, upper_to_lower_var_indices, upper_var_to_lower_ctr)
+
+                if isempty(A_N)
+                    @warn("No bilinear products of lower level dual and primal variables found in upper level objective. Skipping linearization process.")
+                    linearize = false
+                else
+                    AB_N, B, lower_obj_terms, lower_obj_type_handled = 
+                        get_lower_obj_coefs_of_upper_times_lower_primals(lower, lower_var_indices_of_upper_vars, A_N)
+                    if !lower_obj_type_handled
+                        @warn("Linearizing bilinear terms does not handle lower level objective type $(MOI.get(lower, MOI.ObjectiveFunctionType())). Skipping linearization process.")
+                        linearize = false
+                    end
+                end
+            end
+        else
+            @warn("Upper objective must be quadratic and there must lower dual variables in the upper model to linearize bilinear terms. Skipping linearization process.")
```

Comment:
also only for tests right?

Reply:
Meant for users that provide `linearize_bilinear_upper_terms=true` but do not have a model that is appropriate for linearization.

Reply:
I mean, if linearization is active but there is nothing to do, just use @debug, but if it failed to linearized due to some condition not being met, then use @warn

Reply:
Got it, addressed in b02f8bc